sdk/cosmos/azure-cosmos/azure/cosmos/aio/_retry_utility_async.py

```diff
@@ -109,15 +112,24 @@ async def ExecuteAsync(client, global_endpoint_manager, function, *args, **kwarg
                 retry_policy = sessionRetry_policy
             elif exceptions._partition_range_is_gone(e):
                 retry_policy = partition_key_range_gone_retry_policy
+            elif exceptions._container_recreate_exception(e):
+                retry_policy = container_recreate_retry_policy
             elif e.status_code in (StatusCodes.REQUEST_TIMEOUT, e.status_code == StatusCodes.SERVICE_UNAVAILABLE):
                 retry_policy = timeout_failover_retry_policy
             else:
                 retry_policy = defaultRetry_policy
 
+            try:
+                # This supports retry policies that need to be awaited
+                should_retry = await retry_policy.ShouldRetryAsync(e)
+            except AttributeError:
```

Comment:
`ShouldRetry` should not be an async method - all it's doing is analyzing the exception object to determine whether it should be retried. We should not be making any network calls.

Reply:
This is the current dilemma with the container recreate retry. In order to retry we need to refresh the container properties cache which requires a network call, before we retry the operation.

Reply:
By all means we can do that - however that does not belong in the `ShouldRetry` method. This method is simply a boolean to indicate whether that follow-up refresh call _should_ be made or not.
If the answer is always yes, then `ShouldRetry` just returns `True`.